src/test/scala/cosas/DependentFunctionsTests.scala

```diff
@@ -0,0 +1,157 @@
+package ohnosequences.cosas.tests
+
+import ohnosequences.cosas._, fns._
+
+object sampleFunctions {
+
+  case object size extends DepFn1[Any,Int] {
+
+    implicit val sizeForInt: App1[this.type,Int,Int]  = this at { x: Int    => x }
+    implicit val sizeForStr: App1[this.type,String,Int]   = this at { x: String => x.length }
+    implicit val sizeForChar: App1[this.type,Char,Int]  = this at { x: Char   => 1 }
```

Comment:
It's possible, but requires a bit of juggling. Once we would have real products would be a nice addition. Something like an HList type wrapping standard tuples would be great.